diff --git a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/app_srtm.c b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/app_srtm.c
index ffbdfdd..b8e880e 100644
--- a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/app_srtm.c
+++ b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/app_srtm.c
@@ -854,7 +854,7 @@ static void APP_SRTM_DoWakeup(void *param)
             APP_SRTM_ConfigUSBWakeup(false);
         }
         /* Power on regulators which were closed on suspend */
-        PRINTF("CA7 in VLLS status, power on it and wakeup!\r\n");
+        PRINTF("CA7 in VLLS status, power it on and wakeup!\r\n");
         APP_PowerOnCA7(true);
         /* Wake up CA7 */
         MU_TriggerInterrupts(MUA, kMU_NmiInterruptTrigger);
@@ -1308,6 +1308,11 @@ static void APP_SRTM_InitPmicService(void)
     bd70528Config.I2C_ReceiveFunc = PMIC_I2C_ReceiveFunc;
     BD70528_Init(&bd70528Handle, &bd70528Config);
 
+    /* Configure supported voltage ranges */
+    BD70528_WriteReg(&bd70528Handle, BD70528_BUCK3_VOLT_WIN_CONF, 0x88); /* 1.2V-1.2V */
+    BD70528_WriteReg(&bd70528Handle, BD70528_LDO1_VOLT_CONF, 0x03); /* LDO1=1.8V */
+    BD70528_WriteReg(&bd70528Handle, BD70528_LDO1_EN_CONF, 0x00); /* LDO1 disabled */
+
     /* Create and register PMIC service */
     pmicAdapter = SRTM_Bd70528Adapter_Create(&bd70528Handle);
     service = SRTM_PmicService_Create(pmicAdapter);
diff --git a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/app_srtm.h b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/app_srtm.h
index 524d46b..f05b17f 100644
--- a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/app_srtm.h
+++ b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/app_srtm.h
@@ -68,7 +68,8 @@
 #define APP_PIN_BT_HOST_WAKE (0x0107U) /* PTB7 */
 #define APP_PIN_WL_HOST_WAKE (0x001FU) /* PTA31 */
 #define APP_PIN_VOL_MINUS (0x000DU)    /* PTA13 */
-#define APP_PIN_VOL_PLUS (0x0003U)     /* PTA3 */
+#define APP_PIN_VOL_PLUS (0x010EU)     /* PTB14 */
+//#define APP_PIN_VOL_PLUS (0x0017U)     /* PTA23 */
 
 /* Keypad index in Linux OS */
 #define APP_KEYPAD_INDEX_ONOFF (116U)
diff --git a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/iar/power_mode_switch.ewd b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/iar/power_mode_switch.ewd
index d9379fe..d723f06 100644
--- a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/iar/power_mode_switch.ewd
+++ b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/iar/power_mode_switch.ewd
@@ -11,7 +11,7 @@
             <name>C-SPY</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>29</version>
+                <version>31</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>1</debug>
                 <option>
@@ -84,11 +84,11 @@
                 </option>
                 <option>
                     <name>OCDynDriverList</name>
-                    <state>JLINK_ID</state>
+                    <state>CMSISDAP_ID</state>
                 </option>
                 <option>
                     <name>OCLastSavedByProductVersion</name>
-                    <state>8.22.2.15993</state>
+                    <state>8.40.1.21529</state>
                 </option>
                 <option>
                     <name>UseFlashLoader</name>
@@ -182,14 +182,6 @@
                     <name>OCMulticoreNrOfCores</name>
                     <state>1</state>
                 </option>
-                <option>
-                    <name>OCMulticoreMaster</name>
-                    <state>0</state>
-                </option>
-                <option>
-                    <name>OCMulticorePort</name>
-                    <state>53461</state>
-                </option>
                 <option>
                     <name>OCMulticoreWorkspace</name>
                     <state></state>
@@ -214,6 +206,18 @@
                     <name>MassEraseBeforeFlashing</name>
                     <state>0</state>
                 </option>
+                <option>
+                    <name>OCMulticoreNrOfCoresSlave</name>
+                    <state>1</state>
+                </option>
+                <option>
+                    <name>OCMulticoreAMPConfigType</name>
+                    <state>0</state>
+                </option>
+                <option>
+                    <name>OCMulticoreSessionFile</name>
+                    <state></state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -292,7 +296,7 @@
                 <option>
                     <name>CMSISDAPResetList</name>
                     <version>1</version>
-                    <state>5</state>
+                    <state>4</state>
                 </option>
                 <option>
                     <name>CMSISDAPHWResetDuration</name>
@@ -421,7 +425,7 @@
                 </option>
                 <option>
                     <name>OCProbeConfig</name>
-                    <state></state>
+                    <state>$TOOLKIT_DIR$\config\debugger\NXP\iMX7ULP.ProbeConfig</state>
                 </option>
                 <option>
                     <name>CMSISDAPProbeConfigRadio</name>
@@ -999,7 +1003,7 @@
             <name>STLINK_ID</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>5</version>
+                <version>6</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>1</debug>
                 <option>
@@ -1101,7 +1105,7 @@
                 </option>
                 <option>
                     <name>CCSTLinkJtagSpeedList</name>
-                    <version>0</version>
+                    <version>2</version>
                     <state>0</state>
                 </option>
                 <option>
@@ -1112,6 +1116,11 @@
                     <name>CCSTLinkUseServerSelect</name>
                     <state>0</state>
                 </option>
+                <option>
+                    <name>CCSTLinkProbeList</name>
+                    <version>0</version>
+                    <state>1</state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -1214,7 +1223,7 @@
             <name>XDS100_ID</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>6</version>
+                <version>8</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>1</debug>
                 <option>
@@ -1376,6 +1385,18 @@
                     <name>CCXds100SWOPort</name>
                     <state>1</state>
                 </option>
+                <option>
+                    <name>CCXDSTargetVccEnable</name>
+                    <state>0</state>
+                </option>
+                <option>
+                    <name>CCXDSTargetVoltage</name>
+                    <state>###Uninitialized###</state>
+                </option>
+                <option>
+                    <name>OCXDSDigitalStatesConfigFile</name>
+                    <state>1</state>
+                </option>
             </data>
         </settings>
         <debuggerPlugins>
@@ -1395,14 +1416,26 @@
                 <file>$TOOLKIT_DIR$\plugins\rtos\FreeRtos\FreeRtosArmPlugin.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\HWRTOSplugin\HWRTOSplugin.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\Mbed\MbedArmPlugin.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\Mbed\MbedArmPlugin2.ENU.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\OpenRTOS\OpenRTOSPlugin.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\RemedyRtosViewer\RemedyRtosViewer.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\SafeRTOS\SafeRTOSPlugin.ewplugin</file>
                 <loadFlag>0</loadFlag>
@@ -1435,10 +1468,6 @@
                 <file>$TOOLKIT_DIR$\plugins\rtos\uCOS-III\uCOS-III-KA-CSpy.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
-            <plugin>
-                <file>$EW_DIR$\common\plugins\CodeCoverage\CodeCoverage.ENU.ewplugin</file>
-                <loadFlag>1</loadFlag>
-            </plugin>
             <plugin>
                 <file>$EW_DIR$\common\plugins\Orti\Orti.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
@@ -1463,7 +1492,7 @@
             <name>C-SPY</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>29</version>
+                <version>31</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>1</debug>
                 <option>
@@ -1634,14 +1663,6 @@
                     <name>OCMulticoreNrOfCores</name>
                     <state>1</state>
                 </option>
-                <option>
-                    <name>OCMulticoreMaster</name>
-                    <state>0</state>
-                </option>
-                <option>
-                    <name>OCMulticorePort</name>
-                    <state>53461</state>
-                </option>
                 <option>
                     <name>OCMulticoreWorkspace</name>
                     <state></state>
@@ -1666,6 +1687,18 @@
                     <name>MassEraseBeforeFlashing</name>
                     <state>0</state>
                 </option>
+                <option>
+                    <name>OCMulticoreNrOfCoresSlave</name>
+                    <state>1</state>
+                </option>
+                <option>
+                    <name>OCMulticoreAMPConfigType</name>
+                    <state>0</state>
+                </option>
+                <option>
+                    <name>OCMulticoreSessionFile</name>
+                    <state></state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -2451,7 +2484,7 @@
             <name>STLINK_ID</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>5</version>
+                <version>6</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>1</debug>
                 <option>
@@ -2553,7 +2586,7 @@
                 </option>
                 <option>
                     <name>CCSTLinkJtagSpeedList</name>
-                    <version>0</version>
+                    <version>2</version>
                     <state>0</state>
                 </option>
                 <option>
@@ -2564,6 +2597,11 @@
                     <name>CCSTLinkUseServerSelect</name>
                     <state>0</state>
                 </option>
+                <option>
+                    <name>CCSTLinkProbeList</name>
+                    <version>0</version>
+                    <state>1</state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -2666,7 +2704,7 @@
             <name>XDS100_ID</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>6</version>
+                <version>8</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>1</debug>
                 <option>
@@ -2828,6 +2866,18 @@
                     <name>CCXds100SWOPort</name>
                     <state>1</state>
                 </option>
+                <option>
+                    <name>CCXDSTargetVccEnable</name>
+                    <state>0</state>
+                </option>
+                <option>
+                    <name>CCXDSTargetVoltage</name>
+                    <state>###Uninitialized###</state>
+                </option>
+                <option>
+                    <name>OCXDSDigitalStatesConfigFile</name>
+                    <state>1</state>
+                </option>
             </data>
         </settings>
         <debuggerPlugins>
@@ -2847,14 +2897,26 @@
                 <file>$TOOLKIT_DIR$\plugins\rtos\FreeRtos\FreeRtosArmPlugin.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\HWRTOSplugin\HWRTOSplugin.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\Mbed\MbedArmPlugin.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\Mbed\MbedArmPlugin2.ENU.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\OpenRTOS\OpenRTOSPlugin.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\RemedyRtosViewer\RemedyRtosViewer.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\SafeRTOS\SafeRTOSPlugin.ewplugin</file>
                 <loadFlag>0</loadFlag>
@@ -2887,10 +2949,6 @@
                 <file>$TOOLKIT_DIR$\plugins\rtos\uCOS-III\uCOS-III-KA-CSpy.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
-            <plugin>
-                <file>$EW_DIR$\common\plugins\CodeCoverage\CodeCoverage.ENU.ewplugin</file>
-                <loadFlag>1</loadFlag>
-            </plugin>
             <plugin>
                 <file>$EW_DIR$\common\plugins\Orti\Orti.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
@@ -2915,7 +2973,7 @@
             <name>C-SPY</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>29</version>
+                <version>31</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>0</debug>
                 <option>
@@ -3086,14 +3144,6 @@
                     <name>OCMulticoreNrOfCores</name>
                     <state>1</state>
                 </option>
-                <option>
-                    <name>OCMulticoreMaster</name>
-                    <state>0</state>
-                </option>
-                <option>
-                    <name>OCMulticorePort</name>
-                    <state>53461</state>
-                </option>
                 <option>
                     <name>OCMulticoreWorkspace</name>
                     <state></state>
@@ -3118,6 +3168,18 @@
                     <name>MassEraseBeforeFlashing</name>
                     <state>0</state>
                 </option>
+                <option>
+                    <name>OCMulticoreNrOfCoresSlave</name>
+                    <state>1</state>
+                </option>
+                <option>
+                    <name>OCMulticoreAMPConfigType</name>
+                    <state>0</state>
+                </option>
+                <option>
+                    <name>OCMulticoreSessionFile</name>
+                    <state></state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -3903,7 +3965,7 @@
             <name>STLINK_ID</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>5</version>
+                <version>6</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>0</debug>
                 <option>
@@ -4005,7 +4067,7 @@
                 </option>
                 <option>
                     <name>CCSTLinkJtagSpeedList</name>
-                    <version>0</version>
+                    <version>2</version>
                     <state>0</state>
                 </option>
                 <option>
@@ -4016,6 +4078,11 @@
                     <name>CCSTLinkUseServerSelect</name>
                     <state>0</state>
                 </option>
+                <option>
+                    <name>CCSTLinkProbeList</name>
+                    <version>0</version>
+                    <state>1</state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -4118,7 +4185,7 @@
             <name>XDS100_ID</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>6</version>
+                <version>8</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>0</debug>
                 <option>
@@ -4280,6 +4347,18 @@
                     <name>CCXds100SWOPort</name>
                     <state>1</state>
                 </option>
+                <option>
+                    <name>CCXDSTargetVccEnable</name>
+                    <state>0</state>
+                </option>
+                <option>
+                    <name>CCXDSTargetVoltage</name>
+                    <state>###Uninitialized###</state>
+                </option>
+                <option>
+                    <name>OCXDSDigitalStatesConfigFile</name>
+                    <state>1</state>
+                </option>
             </data>
         </settings>
         <debuggerPlugins>
@@ -4299,14 +4378,26 @@
                 <file>$TOOLKIT_DIR$\plugins\rtos\FreeRtos\FreeRtosArmPlugin.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\HWRTOSplugin\HWRTOSplugin.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\Mbed\MbedArmPlugin.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\Mbed\MbedArmPlugin2.ENU.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\OpenRTOS\OpenRTOSPlugin.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\RemedyRtosViewer\RemedyRtosViewer.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\SafeRTOS\SafeRTOSPlugin.ewplugin</file>
                 <loadFlag>0</loadFlag>
@@ -4339,10 +4430,6 @@
                 <file>$TOOLKIT_DIR$\plugins\rtos\uCOS-III\uCOS-III-KA-CSpy.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
-            <plugin>
-                <file>$EW_DIR$\common\plugins\CodeCoverage\CodeCoverage.ENU.ewplugin</file>
-                <loadFlag>1</loadFlag>
-            </plugin>
             <plugin>
                 <file>$EW_DIR$\common\plugins\Orti\Orti.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
@@ -4367,7 +4454,7 @@
             <name>C-SPY</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>29</version>
+                <version>31</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>0</debug>
                 <option>
@@ -4538,14 +4625,6 @@
                     <name>OCMulticoreNrOfCores</name>
                     <state>1</state>
                 </option>
-                <option>
-                    <name>OCMulticoreMaster</name>
-                    <state>0</state>
-                </option>
-                <option>
-                    <name>OCMulticorePort</name>
-                    <state>53461</state>
-                </option>
                 <option>
                     <name>OCMulticoreWorkspace</name>
                     <state></state>
@@ -4570,6 +4649,18 @@
                     <name>MassEraseBeforeFlashing</name>
                     <state>0</state>
                 </option>
+                <option>
+                    <name>OCMulticoreNrOfCoresSlave</name>
+                    <state>1</state>
+                </option>
+                <option>
+                    <name>OCMulticoreAMPConfigType</name>
+                    <state>0</state>
+                </option>
+                <option>
+                    <name>OCMulticoreSessionFile</name>
+                    <state></state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -5355,7 +5446,7 @@
             <name>STLINK_ID</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>5</version>
+                <version>6</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>0</debug>
                 <option>
@@ -5457,7 +5548,7 @@
                 </option>
                 <option>
                     <name>CCSTLinkJtagSpeedList</name>
-                    <version>0</version>
+                    <version>2</version>
                     <state>0</state>
                 </option>
                 <option>
@@ -5468,6 +5559,11 @@
                     <name>CCSTLinkUseServerSelect</name>
                     <state>0</state>
                 </option>
+                <option>
+                    <name>CCSTLinkProbeList</name>
+                    <version>0</version>
+                    <state>1</state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -5570,7 +5666,7 @@
             <name>XDS100_ID</name>
             <archiveVersion>2</archiveVersion>
             <data>
-                <version>6</version>
+                <version>8</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>0</debug>
                 <option>
@@ -5732,6 +5828,18 @@
                     <name>CCXds100SWOPort</name>
                     <state>1</state>
                 </option>
+                <option>
+                    <name>CCXDSTargetVccEnable</name>
+                    <state>0</state>
+                </option>
+                <option>
+                    <name>CCXDSTargetVoltage</name>
+                    <state>###Uninitialized###</state>
+                </option>
+                <option>
+                    <name>OCXDSDigitalStatesConfigFile</name>
+                    <state>1</state>
+                </option>
             </data>
         </settings>
         <debuggerPlugins>
@@ -5751,14 +5859,26 @@
                 <file>$TOOLKIT_DIR$\plugins\rtos\FreeRtos\FreeRtosArmPlugin.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\HWRTOSplugin\HWRTOSplugin.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\Mbed\MbedArmPlugin.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\Mbed\MbedArmPlugin2.ENU.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\OpenRTOS\OpenRTOSPlugin.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
+            <plugin>
+                <file>$TOOLKIT_DIR$\plugins\rtos\RemedyRtosViewer\RemedyRtosViewer.ewplugin</file>
+                <loadFlag>0</loadFlag>
+            </plugin>
             <plugin>
                 <file>$TOOLKIT_DIR$\plugins\rtos\SafeRTOS\SafeRTOSPlugin.ewplugin</file>
                 <loadFlag>0</loadFlag>
@@ -5791,10 +5911,6 @@
                 <file>$TOOLKIT_DIR$\plugins\rtos\uCOS-III\uCOS-III-KA-CSpy.ewplugin</file>
                 <loadFlag>0</loadFlag>
             </plugin>
-            <plugin>
-                <file>$EW_DIR$\common\plugins\CodeCoverage\CodeCoverage.ENU.ewplugin</file>
-                <loadFlag>1</loadFlag>
-            </plugin>
             <plugin>
                 <file>$EW_DIR$\common\plugins\Orti\Orti.ENU.ewplugin</file>
                 <loadFlag>0</loadFlag>
diff --git a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/iar/power_mode_switch.ewp b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/iar/power_mode_switch.ewp
index 45a1923..50f580d 100644
--- a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/iar/power_mode_switch.ewp
+++ b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/iar/power_mode_switch.ewp
@@ -66,7 +66,7 @@
                 </option>
                 <option>
                     <name>OGLastSavedByProductVersion</name>
-                    <state>8.32.2.19370</state>
+                    <state>8.40.1.21529</state>
                 </option>
                 <option>
                     <name>GeneralEnableMisra</name>
@@ -746,7 +746,7 @@
             <name>ILINK</name>
             <archiveVersion>0</archiveVersion>
             <data>
-                <version>22</version>
+                <version>23</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>1</debug>
                 <option>
@@ -1061,12 +1061,28 @@
                 </option>
                 <option>
                     <name>IlinkTrustzoneImportLibraryOut</name>
-                    <state>###Unitialized###</state>
+                    <state>power_mode_switch_import_lib.o</state>
                 </option>
                 <option>
                     <name>OILinkExtraOption</name>
                     <state>1</state>
                 </option>
+                <option>
+                    <name>IlinkRawBinaryFile2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinarySymbol2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinarySegment2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinaryAlign2</name>
+                    <state></state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -1161,7 +1177,7 @@
                 </option>
                 <option>
                     <name>OGLastSavedByProductVersion</name>
-                    <state>7.60.1.11206</state>
+                    <state>8.40.1.21529</state>
                 </option>
                 <option>
                     <name>GeneralEnableMisra</name>
@@ -1841,7 +1857,7 @@
             <name>ILINK</name>
             <archiveVersion>0</archiveVersion>
             <data>
-                <version>22</version>
+                <version>23</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>1</debug>
                 <option>
@@ -2163,6 +2179,22 @@
                     <name>OILinkExtraOption</name>
                     <state>1</state>
                 </option>
+                <option>
+                    <name>IlinkRawBinaryFile2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinarySymbol2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinarySegment2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinaryAlign2</name>
+                    <state></state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -2257,7 +2289,7 @@
                 </option>
                 <option>
                     <name>OGLastSavedByProductVersion</name>
-                    <state>7.60.1.11206</state>
+                    <state>8.40.1.21529</state>
                 </option>
                 <option>
                     <name>GeneralEnableMisra</name>
@@ -2303,7 +2335,7 @@
                 </option>
                 <option>
                     <name>RTConfigPath2</name>
-                    <state>$TOOLKIT_DIR$\INC\c\DLib_Config_Normal.h</state>
+                    <state>$TOOLKIT_DIR$\inc\c\DLib_Config_Normal.h</state>
                 </option>
                 <option>
                     <name>GBECoreSlave</name>
@@ -2389,7 +2421,7 @@
                 </option>
                 <option>
                     <name>DSPExtension</name>
-                    <state>0</state>
+                    <state>1</state>
                 </option>
                 <option>
                     <name>TrustZone</name>
@@ -2558,7 +2590,9 @@
                     <state>$PROJ_DIR$/../../../../../devices/MCIMX7U5</state>
                     <state>$PROJ_DIR$/../../../../../devices/MCIMX7U5/utilities/str</state>
                     <state>$PROJ_DIR$/../../../../../devices/MCIMX7U5/utilities</state>
+                    <state>$PROJ_DIR$/../../../../../devices/MCIMX7U5/utilities/debug_console</state>
                     <state>$PROJ_DIR$/../../../../../components/uart</state>
+                    <state>$PROJ_DIR$/../../../../../components/i2c</state>
                     <state>$PROJ_DIR$/../../../../../components/serial_manager</state>
                     <state>$PROJ_DIR$/../../../../../components/serial_manager</state>
                     <state>$PROJ_DIR$/../../../../../devices/MCIMX7U5/drivers</state>
@@ -2576,6 +2610,8 @@
                     <state>$PROJ_DIR$/../../../../../devices/MCIMX7U5/drivers</state>
                     <state>$PROJ_DIR$/../../../../../components/codec/wm8960</state>
                     <state>$PROJ_DIR$/../../../../../components/codec</state>
+                    <state>$PROJ_DIR$/../../../../../components/codec/i2c</state>
+                    <state>$PROJ_DIR$/../../../../../components/codec/port</state>
                     <state>$PROJ_DIR$/../../../../../components/srtm/include</state>
                     <state>$PROJ_DIR$/../../../../../components/srtm/srtm</state>
                     <state>$PROJ_DIR$/../../../../../middleware/multicore/rpmsg_lite/lib/include/platform/imx7ulp_m4</state>
@@ -2612,7 +2648,7 @@
                 </option>
                 <option>
                     <name>CCOptLevelSlave</name>
-                    <state>1</state>
+                    <state>3</state>
                 </option>
                 <option>
                     <name>CompilerMisraRules98</name>
@@ -2931,7 +2967,7 @@
             <name>ILINK</name>
             <archiveVersion>0</archiveVersion>
             <data>
-                <version>22</version>
+                <version>23</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>0</debug>
                 <option>
@@ -3252,6 +3288,22 @@
                     <name>OILinkExtraOption</name>
                     <state>1</state>
                 </option>
+                <option>
+                    <name>IlinkRawBinaryFile2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinarySymbol2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinarySegment2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinaryAlign2</name>
+                    <state></state>
+                </option>
             </data>
         </settings>
         <settings>
@@ -3346,7 +3398,7 @@
                 </option>
                 <option>
                     <name>OGLastSavedByProductVersion</name>
-                    <state>7.60.1.11206</state>
+                    <state>8.40.1.21529</state>
                 </option>
                 <option>
                     <name>GeneralEnableMisra</name>
@@ -3392,7 +3444,7 @@
                 </option>
                 <option>
                     <name>RTConfigPath2</name>
-                    <state>$TOOLKIT_DIR$\INC\c\DLib_Config_Normal.h</state>
+                    <state>$TOOLKIT_DIR$\inc\c\DLib_Config_Normal.h</state>
                 </option>
                 <option>
                     <name>GBECoreSlave</name>
@@ -3478,7 +3530,7 @@
                 </option>
                 <option>
                     <name>DSPExtension</name>
-                    <state>0</state>
+                    <state>1</state>
                 </option>
                 <option>
                     <name>TrustZone</name>
@@ -3703,7 +3755,7 @@
                 </option>
                 <option>
                     <name>CCOptLevelSlave</name>
-                    <state>1</state>
+                    <state>3</state>
                 </option>
                 <option>
                     <name>CompilerMisraRules98</name>
@@ -4024,7 +4076,7 @@
             <name>ILINK</name>
             <archiveVersion>0</archiveVersion>
             <data>
-                <version>22</version>
+                <version>23</version>
                 <wantNonLocal>1</wantNonLocal>
                 <debug>0</debug>
                 <option>
@@ -4346,6 +4398,22 @@
                     <name>OILinkExtraOption</name>
                     <state>1</state>
                 </option>
+                <option>
+                    <name>IlinkRawBinaryFile2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinarySymbol2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinarySegment2</name>
+                    <state></state>
+                </option>
+                <option>
+                    <name>IlinkRawBinaryAlign2</name>
+                    <state></state>
+                </option>
             </data>
         </settings>
         <settings>
diff --git a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/lpm.c b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/lpm.c
index f4b9354..80efa23 100644
--- a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/lpm.c
+++ b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/lpm.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017-2018, NXP
+ * Copyright 2017-2019, NXP
  * All rights reserved.
  *
  * SPDX-License-Identifier: BSD-3-Clause
@@ -31,6 +31,13 @@
  */
 #define SYSTICK_COUNT_PER_TICK (SYSTICK_SOURCE_CLOCK / configTICK_RATE_HZ)
 
+#define PFD_GATES_MASK                                                                               \
+    (SCG_SPLLPFD_PFD0_CLKGATE_MASK | SCG_SPLLPFD_PFD1_CLKGATE_MASK | SCG_SPLLPFD_PFD2_CLKGATE_MASK | \
+     SCG_SPLLPFD_PFD3_CLKGATE_MASK)
+#define PFD_VALID_MASK                                                                         \
+    (SCG_SPLLPFD_PFD0_VALID_MASK | SCG_SPLLPFD_PFD1_VALID_MASK | SCG_SPLLPFD_PFD2_VALID_MASK | \
+     SCG_SPLLPFD_PFD3_VALID_MASK)
+
 struct _lpm_power_mode_listener
 {
     lpm_power_mode_callback_t callback;
@@ -51,6 +58,14 @@ typedef struct _lpm_nvic_context
 static lpm_power_mode_t s_curMode;
 static lpm_nvic_context_t s_nvicContext;
 static volatile uint32_t s_psp;
+static uint32_t s_usbWakeup;
+static uint32_t s_rccr;
+static uint32_t s_qspipcc;
+static uint32_t s_sosccsr;
+static uint32_t s_firccsr;
+static uint32_t s_sirccsr;
+static uint32_t s_spllcsr;
+static uint32_t s_apllcsr;
 
 static SemaphoreHandle_t s_mutex;
 static lpm_power_mode_listener_t *s_listenerHead;
@@ -58,9 +73,9 @@ static lpm_power_mode_listener_t *s_listenerTail;
 
 #if (defined(__ICCARM__))
 static uint8_t s_suspendMem[0x4000] @"M4SuspendRam";
-#elif(defined(__ARMCC_VERSION))
+#elif (defined(__ARMCC_VERSION))
 static uint8_t s_suspendMem[0x4000] __attribute__((section("M4SuspendRam"), zero_init));
-#elif(defined(__GNUC__))
+#elif (defined(__GNUC__))
 static uint8_t s_suspendMem[0x4000] __attribute__((section("M4SuspendRam,\"aw\",%nobits @")));
 #else
 #error Toolchain not supported.
@@ -70,12 +85,126 @@ static uint8_t s_suspendMem[0x4000] __attribute__((section("M4SuspendRam,\"aw\",
 extern bool LPM_Suspend(void);
 extern void LPM_Resume(void);
 
+AT_QUICKACCESS_SECTION_CODE(extern void BOARD_SetRunMode(
+    SCG_Type *scg, uint32_t scgRunConfig, QuadSPI_Type *qspi, clock_ip_name_t qspiClock, uint32_t qspiClockConfig));
+extern bool BOARD_IsRunOnQSPI(void);
+
 /* FreeRTOS implemented Systick handler. */
 extern void xPortSysTickHandler(void);
 
 /*******************************************************************************
  * Code
  ******************************************************************************/
+static void LPM_DisableSOSC(void)
+{
+    QuadSPI_Type *qspi = BOARD_IsRunOnQSPI() ? QuadSPI0 : NULL;
+
+    /* SOSC can only be disabled when A7 is in VLLS */
+    if (MU_GetOtherCorePowerMode(MUA) != kMU_PowerModeDsm)
+    {
+        return;
+    }
+
+    /* Power down OCOTP. */
+    OCOTP_CTRL->HW_OCOTP_PDN = 1;
+
+    s_rccr = 0;
+    /* Only RUN/VLPR may enter VLLS/LLS/VLPS, and VLPR doesn't use SOSC */
+    if (SMC_GetPowerModeState(MSMC0) == kSMC_PowerStateRun)
+    {
+        s_rccr    = SCG0->RCCR;
+        s_qspipcc = *((volatile uint32_t *)kCLOCK_Qspi);
+        /* Switch CPU and QSPI to FIRC */
+        BOARD_SetRunMode(SCG0, (kSCG_SysClkSrcFirc << SCG_RCCR_SCS_SHIFT) | SCG_RCCR_DIVSLOW(1), qspi, kCLOCK_Qspi,
+                         PCC1_PCC_QSPI_OTFAD_CGC_MASK | PCC1_PCC_QSPI_OTFAD_PCS(3));
+        s_spllcsr = SCG0->SPLLCSR & SCG_SPLLCSR_SPLLEN_MASK;
+        s_apllcsr = SCG0->APLLCSR & SCG_APLLCSR_APLLEN_MASK;
+        /* Disable SPLL. */
+        if (s_spllcsr)
+        {
+            SCG0->SPLLCSR &= ~SCG_SPLLCSR_SPLLEN_MASK;
+        }
+        /* Disable APLL. */
+        if (s_apllcsr)
+        {
+            SCG0->APLLCSR &= ~SCG_APLLCSR_APLLEN_MASK;
+        }
+    }
+    s_sosccsr = SCG0->SOSCCSR & SCG_SOSCCSR_SOSCEN_MASK;
+    /* Disable SOSC. */
+    if (s_sosccsr)
+    {
+        SCG0->SOSCCSR &= ~SCG_SOSCCSR_SOSCEN_MASK;
+    }
+    /* Disable FIRC in VLPS. */
+    s_firccsr = SCG0->FIRCCSR & SCG_FIRCCSR_FIRCLPEN_MASK;
+    if (s_firccsr)
+    {
+        SCG0->FIRCCSR &= ~SCG_FIRCCSR_FIRCLPEN_MASK;
+    }
+    /* Disable SIRC in VLPS. */
+    s_sirccsr = SCG0->SIRCCSR & SCG_SIRCCSR_SIRCLPEN_MASK;
+    if (s_sirccsr)
+    {
+        SCG0->SIRCCSR &= ~SCG_SIRCCSR_SIRCLPEN_MASK;
+    }
+}
+
+static void LPM_EnableSOSC(void)
+{
+    QuadSPI_Type *qspi = BOARD_IsRunOnQSPI() ? QuadSPI0 : NULL;
+
+    /* SOSC can only be disabled when A7 is in VLLS */
+    if (MU_GetOtherCorePowerMode(MUA) != kMU_PowerModeDsm)
+    {
+        return;
+    }
+
+    /* Restore FIRC VLPS Enable */
+    if (s_firccsr)
+    {
+        SCG->FIRCCSR |= SCG_FIRCCSR_FIRCLPEN_MASK;
+    }
+
+    /* Restore SIRC VLPS Enable */
+    if (s_sirccsr)
+    {
+        SCG->SIRCCSR |= SCG_SIRCCSR_SIRCLPEN_MASK;
+    }
+
+    /* Restore SOSC clock */
+    if (s_sosccsr)
+    {
+        SCG->SOSCCSR |= SCG_SOSCCSR_SOSCEN_MASK;
+        while (!(SCG->SOSCCSR & SCG_SOSCCSR_SOSCVLD_MASK))
+        {
+        }
+    }
+
+    if (s_rccr)
+    {
+        /* Restore SPLL */
+        if (s_spllcsr)
+        {
+            SCG0->SPLLCSR |= SCG_SPLLCSR_SPLLEN_MASK;
+            while (!(SCG->SPLLCSR & SCG_SPLLCSR_SPLLVLD_MASK))
+            {
+            }
+        }
+        /* Restore APLL */
+        if (s_apllcsr)
+        {
+            SCG0->APLLCSR |= SCG_APLLCSR_APLLEN_MASK;
+            while (!(SCG->APLLCSR & SCG_APLLCSR_APLLVLD_MASK))
+            {
+            }
+        }
+
+        /* Switching CPU and QSPI clock back */
+        BOARD_SetRunMode(SCG0, s_rccr, qspi, kCLOCK_Qspi, s_qspipcc);
+    }
+}
+
 static uint32_t LPM_EnterTicklessIdle(uint32_t timeoutMilliSec, uint64_t *pCounter)
 {
     uint64_t counter;
@@ -84,8 +213,11 @@ static uint32_t LPM_EnterTicklessIdle(uint32_t timeoutMilliSec, uint64_t *pCount
     uint32_t flag;
     uint32_t timeoutTicks;
     uint32_t countPerTick = SYSTICK_COUNT_PER_TICK;
+    uint32_t sourceClock  = SYSTICK_SOURCE_CLOCK;
     lptmr_config_t lptmrConfig;
 
+    assert(sourceClock != 0U);
+
     if ((uint64_t)timeoutMilliSec * configTICK_RATE_HZ < 2 * 1000)
     {
         /* If timeout < 2 ticks, don't do tickless idle. */
@@ -93,14 +225,14 @@ static uint32_t LPM_EnterTicklessIdle(uint32_t timeoutMilliSec, uint64_t *pCount
     }
 
     maxMS = 0xFFFFU / SYSTICK_TICKLESS_CLOCK * 1000;
-    ms = timeoutMilliSec > maxMS ? maxMS : timeoutMilliSec;
+    ms    = timeoutMilliSec > maxMS ? maxMS : timeoutMilliSec;
 
     /* Calculate the LPTMR counter needed for timeout */
     timeoutTicks = (uint64_t)ms * configTICK_RATE_HZ / 1000;
-    counter = (uint64_t)timeoutTicks * countPerTick;
+    counter      = (uint64_t)timeoutTicks * countPerTick;
 
     expired = LPTMR_GetCurrentTimerCount(SYSTICK_BASE);
-    flag = LPTMR_GetStatusFlags(SYSTICK_BASE);
+    flag    = LPTMR_GetStatusFlags(SYSTICK_BASE);
     LPTMR_Deinit(SYSTICK_BASE); /* Flag cleared and timer stopped. */
     NVIC_ClearPendingIRQ(SYSTICK_IRQn);
     if (flag)
@@ -140,7 +272,7 @@ static uint32_t LPM_EnterTicklessIdle(uint32_t timeoutMilliSec, uint64_t *pCount
     /* Initialize the LPTMR */
     LPTMR_Init(SYSTICK_BASE, &lptmrConfig);
     /* Convert count in systick freq to tickless clock count */
-    LPTMR_SetTimerPeriod(SYSTICK_BASE, counter * SYSTICK_TICKLESS_CLOCK / SYSTICK_SOURCE_CLOCK);
+    LPTMR_SetTimerPeriod(SYSTICK_BASE, counter * SYSTICK_TICKLESS_CLOCK / sourceClock);
     /* Enable timer interrupt */
     LPTMR_EnableInterrupts(SYSTICK_BASE, kLPTMR_TimerInterruptEnable);
     /* Restart timer */
@@ -160,6 +292,8 @@ static void LPM_ExitTicklessIdle(uint32_t timeoutTicks, uint64_t timeoutCounter)
     uint32_t countPerTick = SYSTICK_COUNT_PER_TICK;
     lptmr_config_t lptmrConfig;
 
+    assert(countPerTick != 0U);
+
     /* Convert tickless count to systick count. */
     expired = (uint64_t)(LPTMR_GetCurrentTimerCount(SYSTICK_BASE)) * SYSTICK_SOURCE_CLOCK / SYSTICK_TICKLESS_CLOCK;
     LPTMR_Deinit(SYSTICK_BASE);
@@ -174,9 +308,9 @@ static void LPM_ExitTicklessIdle(uint32_t timeoutTicks, uint64_t timeoutCounter)
     else
     {
         /* remaining counter */
-        counter = timeoutCounter - expired;
+        counter       = timeoutCounter - expired;
         completeTicks = timeoutTicks - (counter - 1) / countPerTick - 1;
-        counter = (counter - 1) % countPerTick + 1;
+        counter       = (counter - 1) % countPerTick + 1;
     }
 
     /* Now reinit Systick with systick clock source. */
@@ -267,7 +401,7 @@ bool LPM_IsTargetModeValid(lpm_power_mode_t targetPowerMode, const char **pError
         case kSMC_PowerStateHsrun:
             if (LPM_PowerModeRun != targetPowerMode)
             {
-                errorMsg = "Current mode is HSRUN, please choose RUN mode as target mode.\r\n";
+                errorMsg  = "Current mode is HSRUN, please choose RUN mode as target mode.\r\n";
                 modeValid = false;
             }
             break;
@@ -275,7 +409,7 @@ bool LPM_IsTargetModeValid(lpm_power_mode_t targetPowerMode, const char **pError
         case kSMC_PowerStateRun:
             if (LPM_PowerModeVlpw == targetPowerMode)
             {
-                errorMsg = "Could not enter VLPW mode from RUN mode.\r\n";
+                errorMsg  = "Could not enter VLPW mode from RUN mode.\r\n";
                 modeValid = false;
             }
             break;
@@ -284,13 +418,13 @@ bool LPM_IsTargetModeValid(lpm_power_mode_t targetPowerMode, const char **pError
             if ((LPM_PowerModeWait == targetPowerMode) || (LPM_PowerModeHsrun == targetPowerMode) ||
                 (LPM_PowerModeStop == targetPowerMode))
             {
-                errorMsg = "Could not enter HSRUN/STOP/WAIT modes from VLPR mode.\r\n";
+                errorMsg  = "Could not enter HSRUN/STOP/WAIT modes from VLPR mode.\r\n";
                 modeValid = false;
             }
             break;
 
         default:
-            errorMsg = "Wrong power state.\r\n";
+            errorMsg  = "Wrong power state.\r\n";
             modeValid = false;
             break;
     }
@@ -303,7 +437,7 @@ bool LPM_IsTargetModeValid(lpm_power_mode_t targetPowerMode, const char **pError
             case LPM_PowerModeLls:
                 if (powerMode != kMU_PowerModeDsm)
                 {
-                    errorMsg = "M4 can enter LLS Mode only when A7 in LLS Mode or VLLS Mode!!!\r\n";
+                    errorMsg  = "M4 can enter LLS Mode only when A7 in LLS Mode or VLLS Mode!!!\r\n";
                     modeValid = false;
                 }
                 break;
@@ -311,7 +445,7 @@ bool LPM_IsTargetModeValid(lpm_power_mode_t targetPowerMode, const char **pError
             case LPM_PowerModeVlls:
                 if (powerMode != kMU_PowerModeDsm)
                 {
-                    errorMsg = "M4 can enter VLLS Mode only when A7 in VLLS Mode!!!\r\n";
+                    errorMsg  = "M4 can enter VLLS Mode only when A7 in VLLS Mode!!!\r\n";
                     modeValid = false;
                 }
                 break;
@@ -328,7 +462,7 @@ bool LPM_IsTargetModeValid(lpm_power_mode_t targetPowerMode, const char **pError
             ((LPM_PowerModeHsrun == targetPowerMode) && (kSMC_PowerStateHsrun == curPowerState)) ||
             ((LPM_PowerModeVlpr == targetPowerMode) && (kSMC_PowerStateVlpr == curPowerState)))
         {
-            errorMsg = "Already in the target power mode.\r\n";
+            errorMsg  = "Already in the target power mode.\r\n";
             modeValid = false;
         }
     }
@@ -425,10 +559,14 @@ bool LPM_WaitForInterrupt(uint32_t timeoutMilliSec)
             status = SMC_SetPowerModeWait(MSMC0);
             break;
         case LPM_PowerModeVlps:
+            LPM_DisableSOSC();
             status = SMC_SetPowerModeVlps(MSMC0);
+            LPM_EnableSOSC();
             break;
         case LPM_PowerModeLls:
+            LPM_DisableSOSC();
             status = SMC_SetPowerModeLls(MSMC0);
+            LPM_EnableSOSC();
             break;
         case LPM_PowerModeVlls:
             if (!LPM_Suspend())
@@ -460,15 +598,15 @@ void LPM_RegisterPowerListener(lpm_power_mode_callback_t callback, void *data)
     assert(l);
 
     l->callback = callback;
-    l->data = data;
-    l->next = NULL;
+    l->data     = data;
+    l->next     = NULL;
 
     xSemaphoreTake(s_mutex, portMAX_DELAY);
 
     if (s_listenerHead)
     {
         s_listenerTail->next = l;
-        s_listenerTail = l;
+        s_listenerTail       = l;
     }
     else
     {
@@ -576,7 +714,7 @@ void LPM_NvicStateSave(void)
     uint32_t irqNum;
 
     irqRegs = (SCnSCB->ICTR & SCnSCB_ICTR_INTLINESNUM_Msk) + 1;
-    irqNum = irqRegs * 32;
+    irqNum  = irqRegs * 32;
 
     s_nvicContext.PriorityGroup = NVIC_GetPriorityGrouping();
 
@@ -590,11 +728,11 @@ void LPM_NvicStateSave(void)
         s_nvicContext.IP[i] = NVIC->IP[i];
     }
 
-    s_nvicContext.SHP[0] = SCB->SHP[0];   /* MemManage */
-    s_nvicContext.SHP[1] = SCB->SHP[1];   /* BusFault */
-    s_nvicContext.SHP[2] = SCB->SHP[2];   /* UsageFault */
-    s_nvicContext.SHP[7] = SCB->SHP[7];   /* SVCall */
-    s_nvicContext.SHP[8] = SCB->SHP[8];   /* DebugMonitor */
+    s_nvicContext.SHP[0]  = SCB->SHP[0];  /* MemManage */
+    s_nvicContext.SHP[1]  = SCB->SHP[1];  /* BusFault */
+    s_nvicContext.SHP[2]  = SCB->SHP[2];  /* UsageFault */
+    s_nvicContext.SHP[7]  = SCB->SHP[7];  /* SVCall */
+    s_nvicContext.SHP[8]  = SCB->SHP[8];  /* DebugMonitor */
     s_nvicContext.SHP[10] = SCB->SHP[10]; /* PendSV */
     s_nvicContext.SHP[11] = SCB->SHP[11]; /* SysTick */
 }
@@ -606,7 +744,7 @@ void LPM_NvicStateRestore(void)
     uint32_t irqNum;
 
     irqRegs = (SCnSCB->ICTR & SCnSCB_ICTR_INTLINESNUM_Msk) + 1;
-    irqNum = irqRegs * 32;
+    irqNum  = irqRegs * 32;
 
     NVIC_SetPriorityGrouping(s_nvicContext.PriorityGroup);
 
@@ -620,11 +758,11 @@ void LPM_NvicStateRestore(void)
         NVIC->IP[i] = s_nvicContext.IP[i];
     }
 
-    SCB->SHP[0] = s_nvicContext.SHP[0];   /* MemManage */
-    SCB->SHP[1] = s_nvicContext.SHP[1];   /* BusFault */
-    SCB->SHP[2] = s_nvicContext.SHP[2];   /* UsageFault */
-    SCB->SHP[7] = s_nvicContext.SHP[7];   /* SVCall */
-    SCB->SHP[8] = s_nvicContext.SHP[8];   /* DebugMonitor */
+    SCB->SHP[0]  = s_nvicContext.SHP[0];  /* MemManage */
+    SCB->SHP[1]  = s_nvicContext.SHP[1];  /* BusFault */
+    SCB->SHP[2]  = s_nvicContext.SHP[2];  /* UsageFault */
+    SCB->SHP[7]  = s_nvicContext.SHP[7];  /* SVCall */
+    SCB->SHP[8]  = s_nvicContext.SHP[8];  /* DebugMonitor */
     SCB->SHP[10] = s_nvicContext.SHP[10]; /* PendSV */
     SCB->SHP[11] = s_nvicContext.SHP[11]; /* SysTick */
 }
@@ -633,18 +771,17 @@ void LPM_SystemSuspend(uint32_t psp)
 {
     uint32_t mask0;
 
+    /* Save USB wakeup setting. */
+    s_usbWakeup = SIM->GPR1 & SIM_GPR1_USB_PHY_WAKEUP_ISO_DISABLE_MASK;
+
     mask0 = SIM_SIM_DGO_CTRL0_WR_ACK_DGO_GP6_MASK | SIM_SIM_DGO_CTRL0_WR_ACK_DGO_GP5_MASK |
             SIM_SIM_DGO_CTRL0_WR_ACK_DGO_GP4_MASK | SIM_SIM_DGO_CTRL0_WR_ACK_DGO_GP3_MASK |
             SIM_SIM_DGO_CTRL0_WR_ACK_DGO_GP2_MASK | SIM_SIM_DGO_CTRL0_WR_ACK_DGO_GP1_MASK;
 
     s_psp = psp; /* Save PSP for resume context */
 
-    /* Save data which will be updated by ROM resume. */
-    memcpy(s_suspendMem, ROM_RESERVED_MEM, sizeof(s_suspendMem));
-    memcpy(ROM_HEADER_MEM, ROM_RESERVED_MEM + sizeof(s_suspendMem), ROM_HEADER_SIZE);
-
     /* Setup VLLS Resume Entry. */
-    SIM->SIM_DGO_GP1 = (uint32_t)LPM_Resume;
+    SIM->SIM_DGO_GP1   = (uint32_t)LPM_Resume;
     SIM->SIM_DGO_CTRL0 = (SIM->SIM_DGO_CTRL0 & ~mask0) | SIM_SIM_DGO_CTRL0_UPDATE_DGO_GP1_MASK;
     /* Wait DGO GP1 updated */
     while ((SIM->SIM_DGO_CTRL0 & SIM_SIM_DGO_CTRL0_WR_ACK_DGO_GP1_MASK) == 0)
@@ -654,7 +791,12 @@ void LPM_SystemSuspend(uint32_t psp)
     SIM->SIM_DGO_CTRL0 =
         (SIM->SIM_DGO_CTRL0 & ~(SIM_SIM_DGO_CTRL0_UPDATE_DGO_GP1_MASK | mask0)) | SIM_SIM_DGO_CTRL0_WR_ACK_DGO_GP1_MASK;
 
+    LPM_DisableSOSC();
+    /* Save data which will be updated by ROM resume. */
+    memcpy(s_suspendMem, ROM_RESERVED_MEM, sizeof(s_suspendMem));
+    memcpy(ROM_HEADER_MEM, ROM_RESERVED_MEM + sizeof(s_suspendMem), ROM_HEADER_SIZE);
     SMC_SetPowerModeVlls(MSMC0);
+    LPM_EnableSOSC();
 }
 
 uint32_t LPM_SystemResume(bool resume)
@@ -668,7 +810,7 @@ uint32_t LPM_SystemResume(bool resume)
     SMC_SetPowerModeProtection(MSMC0, kSMC_AllowPowerModeAll);
 
     /* Clear VLLS Resume Entry. */
-    SIM->SIM_DGO_GP1 = 0U;
+    SIM->SIM_DGO_GP1   = 0U;
     SIM->SIM_DGO_CTRL0 = (SIM->SIM_DGO_CTRL0 & ~mask0) | SIM_SIM_DGO_CTRL0_UPDATE_DGO_GP1_MASK;
     /* Wait DGO GP1 updated */
     while ((SIM->SIM_DGO_CTRL0 & SIM_SIM_DGO_CTRL0_WR_ACK_DGO_GP1_MASK) == 0)
@@ -682,5 +824,7 @@ uint32_t LPM_SystemResume(bool resume)
     memcpy(ROM_RESERVED_MEM, s_suspendMem, sizeof(s_suspendMem));
     memcpy(ROM_RESERVED_MEM + sizeof(s_suspendMem), ROM_HEADER_MEM, ROM_HEADER_SIZE);
 
+    SIM->GPR1 = (SIM->GPR1 & ~SIM_GPR1_USB_PHY_WAKEUP_ISO_DISABLE_MASK) | s_usbWakeup;
+
     return s_psp;
 }
diff --git a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/pin_mux.c b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/pin_mux.c
index 93f9d59..0e7410f 100644
--- a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/pin_mux.c
+++ b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/pin_mux.c
@@ -127,7 +127,7 @@ void BOARD_InitPins(void) {                                /*!< Function assigne
     IOMUXC_SetPinConfig(BOARD_INITPINS_PMIC_INT_PIN_FUNCTION_ID,
                         IOMUXC0_SW_MUX_CTL_PAD_PE_MASK |
                         IOMUXC0_SW_MUX_CTL_PAD_PS_MASK);
-    IOMUXC_SetPinMux(BOARD_INITPINS_RTC_CLK_PIN_FUNCTION_ID, 0U);
+//    IOMUXC_SetPinMux(BOARD_INITPINS_RTC_CLK_PIN_FUNCTION_ID, 0U);
     IOMUXC_SetPinMux(BOARD_INITPINS_I2S1_TX_FS_PIN_FUNCTION_ID, 0U);
     IOMUXC_SetPinConfig(BOARD_INITPINS_I2S1_TX_FS_PIN_FUNCTION_ID,
                         IOMUXC0_SW_MUX_CTL_PAD_PE_MASK);
@@ -153,6 +153,12 @@ void BOARD_InitPins(void) {                                /*!< Function assigne
     IOMUXC_SetPinConfig(BOARD_INITPINS_SEN_INT_PIN_FUNCTION_ID,
                         IOMUXC0_SW_MUX_CTL_PAD_PE_MASK |
                         IOMUXC0_SW_MUX_CTL_PAD_PS_MASK);
+
+    IOMUXC_SetPinMux(BOARD_M4_PERIPOWER_EN_PIN_FUNCTION_ID, 0U);
+    IOMUXC_SetPinConfig(BOARD_M4_PERIPOWER_EN_PIN_FUNCTION_ID,
+                        IOMUXC0_SW_MUX_CTL_PAD_PE_MASK |
+                        IOMUXC0_SW_MUX_CTL_PAD_PS_MASK |
+                        IOMUXC0_SW_MUX_CTL_PAD_OBE_MASK);
 }
 
 
@@ -253,7 +259,7 @@ void BOARD_InitPinsNoQSPI(void) {                          /*!< Function assigne
     IOMUXC_SetPinConfig(BOARD_INITPINSNOQSPI_PMIC_INT_PIN_FUNCTION_ID,
                         IOMUXC0_SW_MUX_CTL_PAD_PE_MASK |
                         IOMUXC0_SW_MUX_CTL_PAD_PS_MASK);
-    IOMUXC_SetPinMux(BOARD_INITPINSNOQSPI_RTC_CLK_PIN_FUNCTION_ID, 0U);
+//    IOMUXC_SetPinMux(BOARD_INITPINSNOQSPI_RTC_CLK_PIN_FUNCTION_ID, 0U);
     IOMUXC_SetPinMux(BOARD_INITPINSNOQSPI_QSPIA_SCLK_PIN_FUNCTION_ID, 0U);
     IOMUXC_SetPinConfig(BOARD_INITPINSNOQSPI_QSPIA_SCLK_PIN_FUNCTION_ID,
                         IOMUXC0_SW_MUX_CTL_PAD_PE_MASK);
@@ -298,6 +304,12 @@ void BOARD_InitPinsNoQSPI(void) {                          /*!< Function assigne
     IOMUXC_SetPinConfig(BOARD_INITPINSNOQSPI_SEN_INT_PIN_FUNCTION_ID,
                         IOMUXC0_SW_MUX_CTL_PAD_PE_MASK |
                         IOMUXC0_SW_MUX_CTL_PAD_PS_MASK);
+
+    IOMUXC_SetPinMux(BOARD_M4_PERIPOWER_EN_PIN_FUNCTION_ID, 0U);
+    IOMUXC_SetPinConfig(BOARD_M4_PERIPOWER_EN_PIN_FUNCTION_ID,
+                        IOMUXC0_SW_MUX_CTL_PAD_PE_MASK |
+                        IOMUXC0_SW_MUX_CTL_PAD_PS_MASK |
+                        IOMUXC0_SW_MUX_CTL_PAD_OBE_MASK);
 }
 
 
diff --git a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/pin_mux.h b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/pin_mux.h
index a508842..ed533f7 100644
--- a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/pin_mux.h
+++ b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/pin_mux.h
@@ -80,14 +80,25 @@
 #define BOARD_INITPINS_I2S0_TXD0_LABEL                               "I2S0_TXD0"   /*!< Label */
 #define BOARD_INITPINS_I2S0_TXD0_NAME                                "I2S0_TXD0"   /*!< Identifier name */
 
-/* PTA3 (number AF14), VOL+ */
+#if 1
+/* PTB14 (number ??), VOL+ */
+#define BOARD_INITPINS_VOL_UP_PERIPHERAL                                     PTB   /*!< Device name: PTB */
+#define BOARD_INITPINS_VOL_UP_SIGNAL                                        port   /*!< PTB signal: port */
+#define BOARD_INITPINS_VOL_UP_CHANNEL                                         14   /*!< PTB port channel: 14 */
+#define BOARD_INITPINS_VOL_UP_PIN_NAME                                     PTB14   /*!< Pin name */
+#define BOARD_INITPINS_VOL_UP_PIN_FUNCTION_ID                 IOMUXC_PTB14_PTB14   /*!< Pin function id */
+#define BOARD_INITPINS_VOL_UP_LABEL                                       "VOL+"   /*!< Label */
+#define BOARD_INITPINS_VOL_UP_NAME                                      "VOL_UP"   /*!< Identifier name */
+#else
+/* PTA23 (number ??), VOL+ */
 #define BOARD_INITPINS_VOL_UP_PERIPHERAL                                     PTA   /*!< Device name: PTA */
-#define BOARD_INITPINS_VOL_UP_SIGNAL                                        port   /*!< PTA signal: port */
-#define BOARD_INITPINS_VOL_UP_CHANNEL                                          3   /*!< PTA port channel: 3 */
-#define BOARD_INITPINS_VOL_UP_PIN_NAME                                      PTA3   /*!< Pin name */
-#define BOARD_INITPINS_VOL_UP_PIN_FUNCTION_ID                   IOMUXC_PTA3_PTA3   /*!< Pin function id */
+#define BOARD_INITPINS_VOL_UP_SIGNAL                                        port   /*!< PTB signal: port */
+#define BOARD_INITPINS_VOL_UP_CHANNEL                                         23   /*!< PTB port channel: 23 */
+#define BOARD_INITPINS_VOL_UP_PIN_NAME                                     PTA23   /*!< Pin name */
+#define BOARD_INITPINS_VOL_UP_PIN_FUNCTION_ID                 IOMUXC_PTA23_PTA23   /*!< Pin function id */
 #define BOARD_INITPINS_VOL_UP_LABEL                                       "VOL+"   /*!< Label */
 #define BOARD_INITPINS_VOL_UP_NAME                                      "VOL_UP"   /*!< Identifier name */
+#endif
 
 /* PTA13 (number AF16), VOL- */
 #define BOARD_INITPINS_VOL_DOWN_PERIPHERAL                                   PTA   /*!< Device name: PTA */
@@ -216,12 +227,12 @@
 #define BOARD_INITPINS_PMIC_INT_NAME                                  "PMIC_INT"   /*!< Identifier name */
 
 /* PTB14 (number AG8), RTC_CLK */
-#define BOARD_INITPINS_RTC_CLK_PERIPHERAL                                    RTC   /*!< Device name: RTC */
-#define BOARD_INITPINS_RTC_CLK_SIGNAL                                 rtc_clkout   /*!< RTC signal: rtc_clkout */
-#define BOARD_INITPINS_RTC_CLK_PIN_NAME                                    PTB14   /*!< Pin name */
-#define BOARD_INITPINS_RTC_CLK_PIN_FUNCTION_ID           IOMUXC_PTB14_RTC_CLKOUT   /*!< Pin function id */
-#define BOARD_INITPINS_RTC_CLK_LABEL                                   "RTC_CLK"   /*!< Label */
-#define BOARD_INITPINS_RTC_CLK_NAME                                    "RTC_CLK"   /*!< Identifier name */
+//#define BOARD_INITPINS_RTC_CLK_PERIPHERAL                                    RTC   /*!< Device name: RTC */
+//#define BOARD_INITPINS_RTC_CLK_SIGNAL                                 rtc_clkout   /*!< RTC signal: rtc_clkout */
+//#define BOARD_INITPINS_RTC_CLK_PIN_NAME                                    PTB14   /*!< Pin name */
+//#define BOARD_INITPINS_RTC_CLK_PIN_FUNCTION_ID           IOMUXC_PTB14_RTC_CLKOUT   /*!< Pin function id */
+//#define BOARD_INITPINS_RTC_CLK_LABEL                                   "RTC_CLK"   /*!< Label */
+//#define BOARD_INITPINS_RTC_CLK_NAME                                    "RTC_CLK"   /*!< Identifier name */
 
 /* PTA25 (number AG20), A7_POW_EN# */
 #define BOARD_INITPINSNOQSPI_A7_POW_EN_PERIPHERAL                            PTA   /*!< Device name: PTA */
@@ -291,13 +302,20 @@
 #define BOARD_INITPINSNOQSPI_I2S0_TXD0_NAME                          "I2S0_TXD0"   /*!< Identifier name */
 
 /* PTA3 (number AF14), VOL+ */
-#define BOARD_INITPINSNOQSPI_VOL_UP_PERIPHERAL                               PTA   /*!< Device name: PTA */
-#define BOARD_INITPINSNOQSPI_VOL_UP_SIGNAL                                  port   /*!< PTA signal: port */
-#define BOARD_INITPINSNOQSPI_VOL_UP_CHANNEL                                    3   /*!< PTA port channel: 3 */
-#define BOARD_INITPINSNOQSPI_VOL_UP_PIN_NAME                                PTA3   /*!< Pin name */
-#define BOARD_INITPINSNOQSPI_VOL_UP_PIN_FUNCTION_ID             IOMUXC_PTA3_PTA3   /*!< Pin function id */
-#define BOARD_INITPINSNOQSPI_VOL_UP_LABEL                                 "VOL+"   /*!< Label */
-#define BOARD_INITPINSNOQSPI_VOL_UP_NAME                                "VOL_UP"   /*!< Identifier name */
+//#define BOARD_INITPINSNOQSPI_VOL_UP_PERIPHERAL                               PTA   /*!< Device name: PTA */
+//#define BOARD_INITPINSNOQSPI_VOL_UP_SIGNAL                                  port   /*!< PTA signal: port */
+//#define BOARD_INITPINSNOQSPI_VOL_UP_CHANNEL                                    3   /*!< PTA port channel: 3 */
+//#define BOARD_INITPINSNOQSPI_VOL_UP_PIN_NAME                                PTA3   /*!< Pin name */
+//#define BOARD_INITPINSNOQSPI_VOL_UP_PIN_FUNCTION_ID             IOMUXC_PTA3_PTA3   /*!< Pin function id */
+//#define BOARD_INITPINSNOQSPI_VOL_UP_LABEL                                 "VOL+"   /*!< Label */
+//#define BOARD_INITPINSNOQSPI_VOL_UP_NAME                                "VOL_UP"   /*!< Identifier name */
+#define BOARD_INITPINSNOQSPI_VOL_UP_PERIPHERAL        BOARD_INITPINS_VOL_UP_PERIPHERAL
+#define BOARD_INITPINSNOQSPI_VOL_UP_SIGNAL            BOARD_INITPINS_VOL_UP_SIGNAL
+#define BOARD_INITPINSNOQSPI_VOL_UP_CHANNEL           BOARD_INITPINS_VOL_UP_CHANNEL
+#define BOARD_INITPINSNOQSPI_VOL_UP_PIN_NAME          BOARD_INITPINS_VOL_UP_PIN_NAME
+#define BOARD_INITPINSNOQSPI_VOL_UP_PIN_FUNCTION_ID   BOARD_INITPINS_VOL_UP_PIN_FUNCTION_ID
+#define BOARD_INITPINSNOQSPI_VOL_UP_LABEL             BOARD_INITPINS_VOL_UP_LABEL
+#define BOARD_INITPINSNOQSPI_VOL_UP_NAME              BOARD_INITPINS_VOL_UP_NAME
 
 /* PTA13 (number AF16), VOL- */
 #define BOARD_INITPINSNOQSPI_VOL_DOWN_PERIPHERAL                             PTA   /*!< Device name: PTA */
@@ -519,6 +537,14 @@
 #define BOARD_I2C_CONFIGUREPINS_I2C3_SDA_LABEL                        "I2C3_SDA"   /*!< Label */
 #define BOARD_I2C_CONFIGUREPINS_I2C3_SDA_NAME                         "I2C3_SDA"   /*!< Identifier name */
 
+#define BOARD_M4_PERIPOWER_EN_PERIPHERAL                                     PTA   /*!< Device name: PTA */
+#define BOARD_M4_PERIPOWER_EN_SIGNAL                                        port   /*!< PTA signal: port */
+#define BOARD_M4_PERIPOWER_EN_CHANNEL                                         21   /*!< PTA port channel: 21 */
+#define BOARD_M4_PERIPOWER_EN_PIN_NAME                                     PTA21   /*!< Pin name */
+#define BOARD_M4_PERIPOWER_EN_PIN_FUNCTION_ID                 IOMUXC_PTA21_PTA21   /*!< Pin function id */
+#define BOARD_M4_PERIPOWER_EN_LABEL                                      "PTA21"   /*!< Label */
+#define BOARD_M4_PERIPOWER_EN_NAME                                       "PTA21"   /*!< Identifier name */
+
 /*!
  * @addtogroup pin_mux
  * @{
diff --git a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/power_mode_switch.c b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/power_mode_switch.c
index 20dc4b4..401bbd2 100644
--- a/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/power_mode_switch.c
+++ b/boards/evkmcimx7ulp/demo_apps/rohm_power_mode_switch/power_mode_switch.c
@@ -53,8 +53,8 @@
 
 #define APP_LPTMR1_IRQ_PRIO (5U)
 
-#define LLWU_WAKEUP_PIN_IDX (1U) /* LLWU_P1 used for VOL+ button */
-#define LLWU_WAKEUP_PIN_TYPE kLLWU_ExternalPinFallingEdge
+#define LLWU_WAKEUP_PIN_IDX (13U) /* LLWU_P13 used for VOL+ button */
+#define LLWU_WAKEUP_PIN_TYPE kLLWU_ExternalPinRisingEdge
 #define APP_WAKEUP_BUTTON_NAME "VOL+"
 
 typedef enum _app_wakeup_source
@@ -205,10 +205,10 @@ static void BOARD_InitClockAndPins(void)
 static void APP_Suspend(void)
 {
     /* Uninit PMIC I2C */
-    IOMUXC_SetPinMux(IOMUXC_PTB12_PTB12, 0U);
-    IOMUXC_SetPinMux(IOMUXC_PTB13_PTB13, 0U);
-    IOMUXC_SetPinConfig(IOMUXC_PTB12_PTB12, IOMUXC0_SW_MUX_CTL_PAD_PE_MASK | IOMUXC0_SW_MUX_CTL_PAD_PS_MASK);
-    IOMUXC_SetPinConfig(IOMUXC_PTB13_PTB13, IOMUXC0_SW_MUX_CTL_PAD_PE_MASK | IOMUXC0_SW_MUX_CTL_PAD_PS_MASK);
+    IOMUXC_SetPinMux(IOMUXC_PTB12_ADC1_CH13A_CMP1_IN0, 0U);
+    IOMUXC_SetPinConfig(IOMUXC_PTB12_ADC1_CH13A_CMP1_IN0, 0U);
+    IOMUXC_SetPinMux(IOMUXC_PTB13_ADC1_CH13B_CMP1_IN1, 0U);
+    IOMUXC_SetPinConfig(IOMUXC_PTB13_ADC1_CH13B_CMP1_IN1, 0U);
 
     /* Save SRTM context */
     APP_SRTM_Suspend();
@@ -245,9 +245,12 @@ void APP_PowerPreSwitchHook(smc_power_state_t originPowerState, lpm_power_mode_t
         /*
          * Set pin for current leakage.
          * Debug console RX pin: Set to pinmux to analog.
-         * Debug console TX pin: Don't need to change.
+         * Debug console TX pin: Set to pinmux to analog.
          */
         IOMUXC_SetPinMux(IOMUXC_PTA19_CMP1_IN3_3V, 0);
+        IOMUXC_SetPinConfig(IOMUXC_PTA19_CMP1_IN3_3V, 0);
+        IOMUXC_SetPinMux(IOMUXC_PTA18_CMP1_IN1_3V, 0);
+        IOMUXC_SetPinConfig(IOMUXC_PTA18_CMP1_IN1_3V, 0);
 
         if (LPM_PowerModeVlls == targetMode)
         {
@@ -265,10 +268,13 @@ void APP_PowerPostSwitchHook(smc_power_state_t originPowerState, lpm_power_mode_
             APP_Resume(result);
         }
         /*
-         * Debug console RX pin is set to disable for current leakage, need to re-configure pinmux.
-         * Debug console TX pin: Don't need to change.
+         * Debug console RX pin was set to disable for current leakage, need to re-configure pinmux.
+         * Debug console TX pin was set to disable for current leakage, need to re-configure pinmux.
          */
-        IOMUXC_SetPinMux(IOMUXC_PTA19_LPUART0_RX, 0);
+        IOMUXC_SetPinMux(IOMUXC_PTA18_LPUART0_TX, 0U);
+        IOMUXC_SetPinConfig(IOMUXC_PTA18_LPUART0_TX, IOMUXC0_SW_MUX_CTL_PAD_PE_MASK | IOMUXC0_SW_MUX_CTL_PAD_PS_MASK);
+        IOMUXC_SetPinMux(IOMUXC_PTA19_LPUART0_RX, 0U);
+        IOMUXC_SetPinConfig(IOMUXC_PTA19_LPUART0_RX, IOMUXC0_SW_MUX_CTL_PAD_PE_MASK | IOMUXC0_SW_MUX_CTL_PAD_PS_MASK);
         APP_InitDebugConsole();
     }
     PRINTF("== Power switch %s ==\r\n", result ? "OK" : "FAIL");
@@ -348,14 +354,24 @@ static void APP_IRQDispatcher(IRQn_Type irq, void *param)
         case SNVS_IRQn:
             break;
         case PCTLA_IRQn:
-            if ((1U << APP_PIN_IDX(APP_PIN_VOL_PLUS)) & PORT_GetPinsInterruptFlags(PORTA))
-            {
-                /* Flag will be cleared by app_srtm.c */
-                xSemaphoreGiveFromISR(s_wakeupSig, NULL);
-                portYIELD_FROM_ISR(pdTRUE);
+            if (APP_GPIO_IDX(APP_PIN_VOL_PLUS) == 0) {
+              if ((1U << APP_PIN_IDX(APP_PIN_VOL_PLUS)) & PORT_GetPinsInterruptFlags(PORTA))
+              {
+                  /* Flag will be cleared by app_srtm.c */
+                  xSemaphoreGiveFromISR(s_wakeupSig, NULL);
+                  portYIELD_FROM_ISR(pdTRUE);
+              }
             }
             break;
         case PCTLB_IRQn:
+            if (APP_GPIO_IDX(APP_PIN_VOL_PLUS) == 1) {
+              if ((1U << APP_PIN_IDX(APP_PIN_VOL_PLUS)) & PORT_GetPinsInterruptFlags(PORTB))
+              {
+                  /* Flag will be cleared by app_srtm.c */
+                  xSemaphoreGiveFromISR(s_wakeupSig, NULL);
+                  portYIELD_FROM_ISR(pdTRUE);
+              }
+            }
             break;
         default:
             break;
@@ -828,6 +844,7 @@ void PowerModeSwitchTask(void *pvParameters)
         PRINTF("Press  L for read BD70528 RTC timer.\r\n");
         PRINTF("Press  M for set BD70528 RTC alarm.\r\n");
         PRINTF("Press  N for read BD70528 RTC alarm.\r\n");
+        PRINTF("Press  O for read VOL+ button.\r\n");
         PRINTF("\r\nWaiting for power mode select..\r\n\r\n");
 
         /* Wait for user response */
@@ -945,6 +962,14 @@ void PowerModeSwitchTask(void *pvParameters)
         {
             APP_ReadPmicRtcAlarm();
         }
+        else if ('O' == ch)
+        {
+            uint8_t gpioIdx = APP_GPIO_IDX(APP_PIN_VOL_PLUS);
+            uint8_t pinIdx  = APP_PIN_IDX(APP_PIN_VOL_PLUS);
+            GPIO_Type *const gpios[] = GPIO_BASE_PTRS;
+            uint8_t value = GPIO_PinRead(gpios[gpioIdx], pinIdx);
+            PRINTF("VOL+ button = %x\r\n", value);
+        }
         else
         {
             PRINTF("Invalid command %c[0x%x]\r\n", ch, ch);
@@ -1016,12 +1041,28 @@ static void WorkingTask(void *pvParameters)
 /*! @brief Main function */
 int main(void)
 {
+    /* Power related. */
+    SMC_SetPowerModeProtection(MSMC0, kSMC_AllowPowerModeAll);
+
     BOARD_InitClockAndPins();
+    APP_InitPMC0();
     APP_InitDebugConsole();
 
     APP_SRTM_Init();
     LPM_Init();
 
+    /* Make sure M4_PERIPOWER_EN is high */
+    {
+        /* Define the init structure for the output pin*/
+        gpio_pin_config_t pin_config = {
+            kGPIO_DigitalOutput, 1,
+        };
+        CLOCK_EnableClock(kCLOCK_Rgpio2p0);
+        CLOCK_EnableClock(kCLOCK_PctlA);
+        GPIO_PinInit(GPIOA, BOARD_M4_PERIPOWER_EN_CHANNEL, &pin_config);
+        GPIO_PinWrite(GPIOA, BOARD_M4_PERIPOWER_EN_CHANNEL, 1);
+    }
+
     s_wakeupSig = xSemaphoreCreateBinary();
 
     xTaskCreate(PowerModeSwitchTask, "Main Task", 512U, NULL, tskIDLE_PRIORITY + 1U, NULL);
diff --git a/components/bd70528/bd70528.c b/components/bd70528/bd70528.c
index 44ab486..0b8d476 100644
--- a/components/bd70528/bd70528.c
+++ b/components/bd70528/bd70528.c
@@ -503,7 +503,7 @@ void BD70528_RTC_ConvertDecimal(rtc_time_t *rtc)
 bool BD70528_RTC_SetTime(bd70528_handle_t *handle, rtc_time_t rtc)
 {
     assert(handle);
-    assert(rtc);
+    //assert(rtc);
 
     BD70528_RTC_ConvertDecimal(&rtc);
 
@@ -543,7 +543,7 @@ rtc_time_t BD70528_RTC_ReadTime(bd70528_handle_t *handle)
 bool BD70528_RTC_SetAlarm(bd70528_handle_t *handle, rtc_time_t rtc)
 {
     assert(handle);
-    assert(rtc);
+    //assert(rtc);
 
     BD70528_RTC_ConvertDecimal(&rtc);
 
