diff --git a/components/bd70528/bd70528.c b/components/bd70528/bd70528.c
index 44ab486..0b8d476 100644
--- a/components/bd70528/bd70528.c
+++ b/components/bd70528/bd70528.c
@@ -503,7 +503,7 @@ void BD70528_RTC_ConvertDecimal(rtc_time_t *rtc)
 bool BD70528_RTC_SetTime(bd70528_handle_t *handle, rtc_time_t rtc)
 {
     assert(handle);
-    assert(rtc);
+    //assert(rtc);
 
     BD70528_RTC_ConvertDecimal(&rtc);
 
@@ -543,7 +543,7 @@ rtc_time_t BD70528_RTC_ReadTime(bd70528_handle_t *handle)
 bool BD70528_RTC_SetAlarm(bd70528_handle_t *handle, rtc_time_t rtc)
 {
     assert(handle);
-    assert(rtc);
+    //assert(rtc);
 
     BD70528_RTC_ConvertDecimal(&rtc);
 
